src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -114,5 +159,117 @@ void GcsJobManager::HandleReportJobError(const rpc::ReportJobErrorRequest &reque
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  };
+  auto status = SubmitJob(request, on_done);
+  if (!status.ok()) {
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  }
+}
+
+Status GcsJobManager::SubmitJob(const ray::rpc::SubmitJobRequest &request,
+                                const ray::gcs::StatusCallback &callback) {
+  auto job_id = JobID::FromBinary(request.job_id());
+
+  RAY_LOG(INFO) << "Submitting job " << job_id;
+
+  auto it = jobs_.find(job_id);
+  if (it != jobs_.end()) {
+    RAY_LOG(ERROR) << "Failed to submit job " << job_id << ", job id conflicts.";
+    std::ostringstream ss;
+    ss << "Job id conflicts: " << job_id;
+    return Status::Invalid(ss.str());
+  }
+
+  // Fill in as much information as possible for the dashboard to
+  // show the job info in SUBMITTED state.
+  auto job_table_data = std::make_shared<rpc::JobTableData>();
+  job_table_data->set_state(rpc::JobTableData::SUBMITTED);
+  job_table_data->set_job_id(request.job_id());
+  job_table_data->set_language(request.language());
+  // Set the job payload (the json submitted from dashboard).
+  job_table_data->set_job_payload(request.job_payload());
+  // Mark the job is submitted, this field determines whether to
+  // initialize the job environement or not.
+  job_table_data->set_is_submitted_from_dashboard(true);
+
+  auto driver_client_id = SelectDriver(*job_table_data);
+  if (driver_client_id.IsNil()) {
+    RAY_LOG(ERROR) << "Failed to submit job " << job_id
+                   << ", can't select a node for driver.";
+    std::ostringstream ss;
+    ss << "Can't select a node for driver, job id: " << job_id;
+    return Status::Invalid(ss.str());
+  }
+
+  job_table_data->set_raylet_id(driver_client_id.Binary());
+  auto maybe_node = gcs_node_manager_->GetAliveNode(driver_client_id);
+  if (maybe_node.has_value()) {
+    auto node = maybe_node.value();
+    // The hostname and ip address of the node are the same as the driver.
+    job_table_data->set_driver_hostname(node->node_manager_hostname());
+    job_table_data->set_driver_ip_address(node->node_manager_address());
+  }
+
+  RAY_LOG(INFO) << "Submitting job, job id = " << job_id << ", config is "
+                << job_table_data->config().DebugString();
+  auto on_done = [this, driver_client_id, job_id, job_table_data,
+                  callback](Status status) {
+    RAY_CHECK(jobs_.emplace(job_id, job_table_data).second);
+    RAY_CHECK_OK(gcs_pub_sub_->Publish(JOB_CHANNEL, job_id.Hex(),
+                                       job_table_data->SerializeAsString(), nullptr));
+    driver_node_to_jobs_[driver_client_id].emplace(job_id);
+    if (callback) {
+      callback(status);
+    }
+    RAY_LOG(INFO) << "Finished submitting job, job id = " << job_id;
+  };
+  return gcs_table_storage_->JobTable().Put(job_id, *job_table_data, on_done);
+}
+
+NodeID GcsJobManager::SelectDriver(const rpc::JobTableData &job_data) const {
+  std::vector<NodeID> alive_nodes;
```

Comment:
Minor, no need to copy all the ids. you can just random an index, and iterate that many times on the map.

Reply:
Good suggestion.